src/lib/render-lit-html.ts

```diff
@@ -198,181 +383,107 @@ export function* renderTemplateResult(
-            yield attributeString + '"';
           }
-          skipTo(attrEndOffset);
-          boundAttrsCount += 1;
-          templatePartIndex++;
+          // Property should be reflected to attribute
+          const reflectedName = reflectedAttributeName(
+            op.tagName,
+            propertyName
+          );
+          if (reflectedName !== undefined) {
+            yield `${reflectedName}="${value}"`;
+          }
+        } else if (prefix === '@') {
+          // Event binding, do nothing with values
+        } else if (prefix === '?') {
+          // Boolean attribute binding
+          attributeName = attributeName.substring(1);
+          if (statics.length !== 2 || statics[0] !== '' || statics[1] !== '') {
+            throw new Error(
+              'Boolean attributes can only contain a single expression'
+            );
+          }
+          const value = result.values[partIndex];
+          if (value) {
+            yield attributeName;
+          }
+        } else {
+          const attributeString = `${attributeName}="${getAttrValue(
+            statics,
+            result,
+            partIndex
+          )}"`;
+          // attr.value has the raw attribute value, which may contain multiple
+          // bindings. Replace the markers with their resolved values.
+          yield attributeString;
         }
+        partIndex += statics.length - 1;
+        break;
       }
-
-      if (writeTag || boundAttrsCount > 0) {
-        yield flushTo(node.sourceCodeLocation!.startTag.endOffset);
-      }
-
-      if (boundAttrsCount > 0) {
-        const templatePart = templateParts[templatePartIndex];
-        // templatePart.index is the depth-first node index of the parent node
-        // of this comment.
-        yield `<!--lit-bindings ${templatePart.index}-->`;
+      case 'custom-element-open': {
+        const ctor = op.ctor;
+        // Instantiate the element and stream its render() result
+        let instance: HTMLElement | undefined = undefined;
+        try {
+          instance = new ctor();
+          // const renderer = new LitElementRenderer();
+          // yield* renderer.renderElement(instance as LitElement, renderInfo);
+        } catch (e) {
+          console.error('Exception in custom element constructor', e);
+        }
+        renderInfo.customElementInstanceStack.push(instance);
+        break;
       }
-
-      if (instance !== undefined) {
-        // TODO: look up a renderer instead of creating one
-        const renderer = new LitElementRenderer();
-        yield* renderer.renderElement(instance as LitElement, renderInfo);
+      case 'custom-element-render': {
+        const instance =
+          renderInfo.customElementInstanceStack[
+            renderInfo.customElementInstanceStack.length - 1
+          ];
+        if (instance !== undefined) {
+          const renderer = new LitElementRenderer();
+          yield* renderer.renderElement(instance as LitElement, renderInfo);
```

Comment:
Don't have that yet - there's a long standing issue for it: https://github.com/PolymerLabs/lit-ssr/issues/10

When we instantiate the element and set properties on it, we still haven't closed the start tag, so we'll have the opportunity to write reflected attributes (we might have to defer writing the user-set attributes). But first we need a way to get reflected attributes out of an element - probably by implementing `setAttribute()` on our HTMLElement shim.

Reply:
Might want to consider something like `renderer.renderAttributes(instance, renderInfo)` on the custom element renderer type, to avoid the need to actually run element lifecycle to get to the point where the update/reflected attributes code would call the shimmed `setAttribute` shim.  It's another case where we either need to go more or less in on re-using client-side code.

Reply:
I left a very similar comment here: https://github.com/PolymerLabs/lit-ssr/pull/54#discussion_r427515547 Seems like a good addition to ElementRenderer cc @sorvell

Reply:
That's the opposite direction though.  There's (a) outer template setting attributes on the custom element (which may affect its rendering), and there's (b) the custom element wanting to set attributes on itself as part of its rendering.

I was proposing `renderer.renderAttributes` to cover (b), which would return/emit a string  as part of handling the custom-element open opcode.  https://github.com/PolymerLabs/lit-ssr/pull/54#discussion_r427515547 is needed for (a).